### ISSUE TITLE
Question about implementing gini index as a gating value

### ISSUE THREAD

**Issue by yoom618** [YunshiWen/InterpretGatedNetwork#1]
https://github.com/YunshiWen/InterpretGatedNetwork/blob/5ea6045a1f724a536542080acaf3de3719bc4dbb/models/InterpGN.py#L47-L52

I have a question about the code referred above.

So, if I understand correctly,
since `gini` index will move around [1/c, 1] you guys had normalized them to [0, 1] in `sbm_util`.
thus, `sbm_util` will be close to 0 when the logit value is ambiguous, 
and it will be close to 1 when the corresponding logit is like one-hot vector.

and you made mask with the value as `(sbm_util > gating_value)`.

So i have question about the gating value.

From the result csv that you uploaded, it seems like gating_value = 1 is the best.

But since `sbm_util` cannot be bigger than 1 (if i understand correctly)
it seems like not using SBM and just using FCN network is the best.

I don't think that's what you guys intended, so i want to know where i miss the path.

**Comment by Chao0511** [comment]
Inspired by the question from yoom618, I noticed that in your paper's Table 3: Default hyperparameters, the inference gating value is also set to 1; and in Figure 21: Ablation on η, the accuracy is at its peak when η = 1. Detailed explanation would be appreciated, thank you in advance.

**Comment by YunshiWen** [comment]
Thanks for your questions. Let me first make some clarifications:
1. The prediction of the InterpGN is a mixture of the predictions from the SBM and the DNN, with the mixture ratio of $\eta$ (i.e. `sbm_util`). Therefore, only when $\eta=0$, the prediction is "not using the SBM and just using the FCN", and only when $\eta=1$, the prediction is "fully based on the SBM". For any $0< \eta < 1$, the prediction is the mixture of both experts.
2. Some may argue that this mixture will lose the interpretability from the SBM, as one reviewer pointed out. Therefore, we introduce the gating value during inference to enforce interpretability for samples with $\eta$ greater than the gating value. For such samples, full interpretability is preserved.
3. Note that our argument is that "samples with high $\eta$ are mostly interpretable" since their predictions can be considered as "mostly from the SBM with corrections from the DNN". In such samples, the predictions of the SBM should be identical or at least similar to the predictions of the InterpGN, and therefore interpretability are mostly preserved. This is supported by Figure 17 and Figure 21 in the paper.

In experiments, we observed that:
1. It is possible, although very unlikely, that the SBM is being falsely confident (due to the distribution shift between the training and test data). In some of such cases, the DNN could provide a correction. Therefore, we can see that the gating value 1 gives the best overall performance by always taking the mixture of both experts.
2. Figure 21 shows that using a smaller gating value to enforce interpretability does not significantly degrade the performance. For example, the gating value of 0.8, 0.9, and 1.0 provide almost the same overall accuracy, and the difference between gating value 0.5 and 1.0 is only 0.0035. 
3. In practice, one may adjust the inference gating value based on the actual demand. For example, if precise interpretability is preferred, then user can choose the gating values like 0.6, 0.7, or 0.8. If the mixture of predictions is sufficient, the user can choose the gating values of 0.9 and 1.0 for the optimal performance.

To highlight the answers to your questions:
1. The gating value is a threshold to enforce interpretability. It ensures that samples with $\eta$ greater than the gating value are interpretable, but it does not indicate that samples with $\eta$ smaller than the gating value is "just using the FCN" and not interpretable at all.
2. We choose the default gating value $\underline{\eta}=1$ since we think that the mixture of predictions is still interpretable as long as the SBM utility $\eta$ is high. Figure 21 shows that reducing the gating value to enforce interpretability still offers similar performance.

Hope these could answer your questions, and please let me know if you have any further questions.

**Comment by yoom618** [comment]
Thanks for the detailed answer for the entire process and the results!
I also agree with your viewpoint about the SBM and DNN's relationship.


However, i just want to know about the implementation of `sbm_util` variable.

The case of only using SBM will be when the original `sbm_util` mask is a matrix of ones,
since `output = sbm_util * sbm_out + deep_util * deep_out` should become `output = sbm_out`.

And this can be happen when `sbm_util >= 0` since it has the range btw [0,1].

Thus i just made a issue to know that my understanding about the code is in wrong direction.

**Comment by YunshiWen** [comment]
The `sbm_util` variable in the code is the $\eta(x_i)$ in the paper, which is aways between 0 and 1.

For the time series $x_i$:
- If `sbm_util[i]=1`, then `output[i] = sbm_out[i]`.
- If `sbm_util[i]=0`, then `output[i] = deep_out[i]`.
- If `0 < sbm_util[i] < 1`, `output[i]` is a mixture of both experts.
- When the gating value is given, `sbm_util[i]=1 if sbm_util[i]>gating_value`.

**Comment by yoom618** [comment]
As you mentioned "When the gating value is given, `sbm_util[i]=1 if sbm_util[i]>gating_value`",
I thought the given `gating_value` should be set to 0 if we try to force `sbm_util` mask to become a matrix of ones.
Or making <code>sbm_util <b><</b> gating_value</code> to reverse the result.


So my understanding is like this picture:
<img width="795" alt="Image" src="https://github.com/user-attachments/assets/7c550dcd-bcbd-4f3b-a94c-553a75c4ca6b" />


I don't want to make offense about the paper and your idea because i think it's really an excellent choice to use gini for adjusting the importance of two models in training phase.
(and as you mentioned from the first answer, the performance difference isn't that big regardless of gating_value)

I just thought that your model performance could be higher than the reported results.
(Because it wasn't experimented under the condition of gating value in [0,0.5))

---
You know,,, only using FCN cannot make this performance.

if my thinking is right, then we can see that using SBM in training phase help to enhance the overall performance of both FCN and SBM.
(In other words, I feels like "Even if it's a FCN, if you learn properly, you can produce desired results.")

**Comment by YunshiWen** [comment]
I think there might be a misunderstanding from the figure you draw. On the left side of the gating value, it should be "Use DNN and the SBM (i.e., the Mixture)" instead of "Use DNN".

From the code perspective, we have `sbm_util[i]=1 if sbm_util[i]>gating_value`, while `sbm_util[i]` is unchanged if `sbm_util[i] <= gating_value`.

The gating value is a threshold to enforce interpretability for samples with high `sbm_util`. It is not a value to "make binary assignment to the experts". For a sample with `sbm_util` below the gating value, the prediction is still `output = sbm_util * sbm_out + (1 - sbm_util) * deep_out`, which is a mixture instead of only using the FCN (as you mentioned, a single FCN should perform worse than this mixture).

Using this mixture to enhance the overall performance of both the SBM and the FCN is exactly the main idea of our paper (and MoE in general). But note that the predictions are also mixtures of outputs from the experts instead of having only one expert being activated.

**Comment by yoom618** [comment]
now i get it!
so i totally misunderstood about the `sbm_out` itself..!
such a stupid question it was haha

i thought sbm_out is made of single SBM. i should've read the InterpretGN class code more precisely.

thank you for your answers and 
you patience:)

**Comment by YunshiWen** [comment]
Glad to know your confusion is addressed! Thanks for your question and interest.
